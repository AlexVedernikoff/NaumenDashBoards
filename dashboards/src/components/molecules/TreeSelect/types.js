// @flow
import type {OnSelectEvent, SelectValue} from 'components/types';

type Options = {
	[string]: Object
};

export type OnChangeLabelEvent = {
	label: string,
	name: string
};

export type OnRemoveEvent = {
	name: string
};

export type Props = {
	getOptionLabel?: (option: SelectValue) => string,
	getOptionValue?: (option: SelectValue) => any,
	initialSelected: Array<string>,
	name: string,
	onChangeLabel: OnChangeLabelEvent => void,
	onRemove: OnRemoveEvent => void,
	onSelect: OnSelectEvent => void | Promise<void>,
	options: Options,
	placeholder: string,
	removable: boolean,
	value: SelectValue | null
};

export type State = {
	showForm: boolean,
	showList: boolean
};
