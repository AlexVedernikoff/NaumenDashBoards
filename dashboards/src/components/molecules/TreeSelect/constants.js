// @flow
import Container from 'components/atoms/Container';
import Node from './components/Node';
import SearchInput from 'components/atoms/SearchInput';
import Tree from './components/Tree';

const DEFAULT_COMPONENTS = {
	IndicatorsContainer: Container,
	LabelContainer: Container,
	MenuContainer: Container,
	Node,
	SearchInput,
	Tree,
	ValueContainer: Container
};

export {
	DEFAULT_COMPONENTS
};
