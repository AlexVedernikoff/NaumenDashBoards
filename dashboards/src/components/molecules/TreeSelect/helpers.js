// @flow
import type {Option} from './types';

/**
 * Возвращает представление опции
 * @param {Option} option - опция
 * @returns {string}
 */
const getOptionLabel = (option: Option | null): string => option?.label ?? '';

/**
 * Возвращает значение опции
 * @param {Option} option - опция
 * @returns {?string}
 */
const getOptionValue = (option: Option | null): ?string => option?.value;

export {
	getOptionLabel,
	getOptionValue
};
