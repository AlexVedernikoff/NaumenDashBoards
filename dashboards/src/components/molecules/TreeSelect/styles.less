@import '/styles/variables';

:local {
	@height: 28px;

	.container {
		height: @height;
		padding: 5px 0;
	}

	.select {
		align-items: center;
		border-bottom: 2px solid @black;
		box-sizing: border-box;
		cursor: pointer;
		display: flex;
		height: 100%;
		position: relative;

		&:hover {
			.indicators {
				visibility: visible;
			}
		}
	}

	.indicators {
		align-items: center;
		display: flex;
		margin-left: auto;
		visibility: hidden;
	}

	.list {
		left: 0;
		position: absolute;
		right: 0;
		top: @height;
		z-index: 1;
	}

	.valueContainer {
		flex: auto;
		overflow: hidden;
		text-overflow: ellipsis;
		white-space: nowrap;
	}

	.value {
		align-items: center;
		display: flex;
	}

	.label {
		overflow: hidden;
		text-overflow: ellipsis;
	}

	.placeholder {
		align-items: center;
		display: flex;
		height: 24px;
		color: @darkgrey;
	}
}
