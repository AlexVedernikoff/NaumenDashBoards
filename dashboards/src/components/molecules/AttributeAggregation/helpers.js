// @flow
import {ATTRIBUTE_SETS} from 'store/sources/attributes/constants';
import {DEFAULT_AGGREGATION_OPTIONS, INTEGER_AGGREGATION_OPTIONS} from './constants';

const getAggregationOptions = (attribute: Object | null) => {
	return attribute && ATTRIBUTE_SETS.NUMBER.includes(attribute.type)
		? [...INTEGER_AGGREGATION_OPTIONS, ...DEFAULT_AGGREGATION_OPTIONS]
		: DEFAULT_AGGREGATION_OPTIONS;
};

const getAggregationLabel = (aggregation: string) => {
	const value = [...INTEGER_AGGREGATION_OPTIONS, ...DEFAULT_AGGREGATION_OPTIONS].find(o => o.value === aggregation);
	return value ? value.label : aggregation;
};

const getDefaultAggregation = (attribute: Object | null) => getAggregationOptions(attribute)[0].value;

export {
	getAggregationLabel,
	getAggregationOptions,
	getDefaultAggregation
};
