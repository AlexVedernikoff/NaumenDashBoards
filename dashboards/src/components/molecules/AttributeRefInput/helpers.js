// @flow
import type {Attribute} from 'store/sources/attributes/types';
import {DATETIME_GROUP_OPTIONS, DEFAULT_AGGREGATION_OPTIONS, DEFAULT_GROUP_OPTIONS, INTEGER_AGGREGATION_OPTIONS} from './constants';
import {TYPES} from 'store/sources/attributes/constants';

const getAggregateOptions = (attribute: Attribute | null) => {
	return attribute && TYPES.INTEGER.includes(attribute.type)
		? [...INTEGER_AGGREGATION_OPTIONS, ...DEFAULT_AGGREGATION_OPTIONS]
		: DEFAULT_AGGREGATION_OPTIONS;
};

const getGroupOptions = (attribute: Attribute | null) => {
	if (attribute && TYPES.DATE.includes(attribute.type)) {
		return DATETIME_GROUP_OPTIONS;
	}

	return DEFAULT_GROUP_OPTIONS;
};

export {
	getAggregateOptions,
	getGroupOptions
};
