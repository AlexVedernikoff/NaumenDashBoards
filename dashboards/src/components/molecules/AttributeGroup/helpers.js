// @flow
import {createDefaultGroup} from 'store/widgets/helpers';
import {DATETIME_SYSTEM_GROUP, DEFAULT_SYSTEM_GROUP} from 'store/widgets/constants';
import {DATETIME_SYSTEM_OPTIONS, DEFAULT_SYSTEM_OPTIONS} from './constants';
import {TYPES} from 'store/sources/attributes/constants';

const getSystemGroupOptions = (attribute: Object | null) => {
	if (attribute && TYPES.DATE.includes(attribute.type)) {
		return DATETIME_SYSTEM_OPTIONS;
	}

	return DEFAULT_SYSTEM_OPTIONS;
};

const getDefaultSystemGroup = (attribute: Object) => TYPES.DATE.includes(attribute.type)
		? createDefaultGroup(DATETIME_SYSTEM_GROUP.MONTH)
		: createDefaultGroup(DEFAULT_SYSTEM_GROUP.OVERLAP);

export {
	getDefaultSystemGroup,
	getSystemGroupOptions
};
