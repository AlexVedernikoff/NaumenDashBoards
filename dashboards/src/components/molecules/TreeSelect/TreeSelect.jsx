// @flow
import Container from 'src/components/atoms/Container';
import {debounce} from 'src/helpers';
import IconButton from 'components/atoms/IconButton';
import {ICON_NAMES} from 'components/atoms/Icon';
import LabelEditingForm from 'components/molecules/InputForm';
import type {Node} from 'components/molecules/MaterialTreeSelect/components/Tree/types';
import OutsideClickDetector from 'components/atoms/OutsideClickDetector';
import type {Props, State} from './types';
import React, {PureComponent} from 'react';
import SearchInput from 'src/components/atoms/SearchInput';
import styles from './styles.less';
import Tree from 'components/molecules/MaterialTreeSelect/components/Tree';

export class TreeSelect extends PureComponent<Props, State> {
	static defaultProps = {
		initialSelected: [],
		name: '',
		placeholder: 'Выберите значение'
	};

	state = {
		searchValue: '',
		showForm: false,
		showMenu: false
	};

	getOptionLabel = (option: Object) => {
		const {getOptionLabel} = this.props;
		let label = '';

		if (option) {
			label = getOptionLabel ? getOptionLabel(option) : option.label;
		}

		return label;
	};

	getOptionValue = (option: Object) => {
		const {getOptionValue} = this.props;
		let value = '';

		if (option) {
			value = getOptionValue ? getOptionValue(option) : option.value;
		}

		return value;
	};

	handleChangeLabel = (label: string) => {
		const {name, onChangeLabel} = this.props;

		this.hideForm();
		onChangeLabel({label, name});
	};

	handleChangeSearchInput = (searchValue: string) => this.setState({searchValue});

	handleClick = () => this.setState({showMenu: true});

	handleClickIndicators = (e: Event) => e.stopPropagation();

	handleRemoveValue = () => {
		const {name, onSelect} = this.props;

		onSelect({name, value: null});
	};

	handleSelect = (node: Node) => {
		const {name, onSelect} = this.props;

		this.setState({showMenu: false});
		onSelect({name, value: node.value});
	};

	handleToggleList = () => this.setState({showMenu: !this.state.showMenu});

	hideForm = () => this.setState({showForm: false});

	hideTree = () => this.setState({showMenu: false});

	showForm = () => this.setState({showForm: true});

	renderIndicators = () => {
		const {value} = this.props;

		if (value) {
			return (
				<div className={styles.indicators} onClick={this.handleClickIndicators}>
					<IconButton icon={ICON_NAMES.EDIT} onClick={this.showForm} />
					<IconButton icon={ICON_NAMES.REMOVE} onClick={this.handleRemoveValue} />
				</div>
			);
		}
	};

	renderLabel = () => {
		const label = this.getOptionLabel(this.props.value);

		return <div className={styles.label} title={label}>{label}</div>;
	};

	renderLabelEditingForm = () => {
		const {value} = this.props;
		const {showForm} = this.state;

		if (showForm) {
			const label = this.getOptionLabel(value);

			return (
				<LabelEditingForm
					onClose={this.hideForm}
					onSubmit={this.handleChangeLabel}
					value={label}
				/>
			);
		}
	};

	renderList = () => {
		const {initialSelected, options, value} = this.props;
		const {searchValue} = this.state;

		return (
			<Tree
				getOptionLabel={this.getOptionLabel}
				getOptionValue={this.getOptionValue}
				initialSelected={initialSelected}
				onSelect={this.handleSelect}
				options={options}
				searchValue={searchValue}
				value={value}
			/>
		);
	};

	renderMenu = () => {
		const {showMenu} = this.state;

		if (showMenu) {
			return (
				<Container className={styles.menu}>
					{this.renderSearchInput()}
					{this.renderList()}
				</Container>
			);
		}
	};

	renderPlaceholder = () => (
		<div className={styles.placeholder}>
			{this.props.placeholder}
		</div>
	);

	renderSearchInput = () => {
		const {searchValue} = this.state;

		return <SearchInput onChange={debounce(this.handleChangeSearchInput, 500)} value={searchValue} />;
	};

	renderSelect = () => (
		<OutsideClickDetector onClickOutside={this.hideTree}>
			<div className={styles.select}>
				{this.renderValueContainer()}
				{this.renderMenu()}
			</div>
		</OutsideClickDetector>
	);

	renderValue = () => (
		<div className={styles.value}>
			{this.renderLabel()}
			{this.renderIndicators()}
		</div>
	);

	renderValueContainer = () => {
		const {value} = this.props;
		const content = value ? this.renderValue() : this.renderPlaceholder();

		return <div className={styles.valueContainer} onClick={this.handleToggleList}>{content}</div>;
	};

	render () {
		const content = this.state.showForm ? this.renderLabelEditingForm() : this.renderSelect();

		return (
			<div className={styles.container}>
				{content}
			</div>
		);
	}
}

export default TreeSelect;
