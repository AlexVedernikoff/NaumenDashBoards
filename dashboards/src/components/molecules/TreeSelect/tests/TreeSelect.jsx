import {getNode} from 'components/molecules/TreeSelect/components/Node/tests/helpers';
import {getProps} from './helpers';
import {getOptionLabel, getOptionValue} from 'components/molecules/TreeSelect/helpers';
import {getTree} from 'components/molecules/TreeSelect/components/Tree/tests/helpers';
import React from 'react';
import {shallow} from 'enzyme';
import TreeSelect from 'components/molecules/TreeSelect';

describe('TreeSelect', () => {
	it('renders when no options provided', () => {
		const wrapper = shallow(<TreeSelect {...getProps()} options={{}} />);

		expect(wrapper).toMatchSnapshot();
	});

	it('renders menu when "showMenu" property is changing', () => {
		const wrapper = shallow(<TreeSelect {...getProps()} options={getTree(2, 2)} />);

		expect(wrapper).toMatchSnapshot();

		wrapper.setState({showMenu: true});

		expect(wrapper).toMatchSnapshot();
	});

	it('renders when "removable" property is set to "true"', () => {
		const wrapper = shallow(<TreeSelect {...getProps()} options={{}} removable={true} value={{}} />);

		expect(wrapper).toMatchSnapshot();
	});

	it('handles onSelect properly and calls provided callback', () => {
		const mockOnSelect = jest.fn();
		const props = getProps();
		const node = getNode();

		const wrapper = shallow(<TreeSelect {...props} onSelect={mockOnSelect} />);

		wrapper.instance().handleSelect(node);

		expect(mockOnSelect.mock.calls.length).toEqual(1);
		expect(mockOnSelect.mock.calls[0][0]).toEqual({
			name: props.name,
			value: node
		});
	});

	it('handles onRemove and calls provided callback', () => {
		const mockOnRemove = jest.fn();
		const props = getProps();

		const wrapper = shallow(<TreeSelect {...props} onRemove={mockOnRemove} />);

		wrapper.instance().handleClickRemoveButton();

		expect(mockOnRemove.mock.calls.length).toEqual(1);
		expect(mockOnRemove.mock.calls[0][0]).toEqual(props.name);
	});

	it('method "getOptionLabel" returns label of "value" property or node value', () => {
		const label = 'test';

		expect(getOptionLabel({label})).toEqual(label);
		expect(getOptionLabel(null)).toEqual('');
	});

	it('method "getOptionValue" returns value of "value" property or node value', () => {
		const value = 'test';

		expect(getOptionValue({value})).toEqual(value);
		expect(getOptionValue(null)).toEqual(undefined);
	});

	it('method "getNodeLabel" returns node label', () => {
		const node = getNode();
		const wrapper = shallow(<TreeSelect {...getProps()} />);

		expect(wrapper.instance().getNodeLabel(node)).toEqual(node.value.label);
	});

	it('method "getNodeValue" returns node value', () => {
		const node = getNode();
		const wrapper = shallow(<TreeSelect {...getProps()} />);

		expect(wrapper.instance().getNodeValue(node)).toEqual(node.value.value);
	});

	it('method "getFoundTree" returns found tree when uses one level of the tree', () => {
		const options = getTree(1, 3);
		const wrapper = shallow(<TreeSelect {...getProps()} options={options} />);
		const fountOptions = wrapper.instance().getFoundTree(options, 'node1');

		expect(Object.keys(fountOptions).length).toEqual(1);
	});

	it('method "getFoundTree" returns found tree when uses nested tree', () => {
		const options = getTree(2, 3);
		const wrapper = shallow(<TreeSelect {...getProps()} options={options} />);
		const fountTree = wrapper.instance().getFoundTree(options, 'node1');

		expect(Object.keys(fountTree).length).toEqual(6);
	});

	it('method "getFoundTree" returns found tree is not case sensitive', () => {
		const options = getTree(2, 3);
		const wrapper = shallow(<TreeSelect {...getProps()} options={options} />);
		const upperFoundTree = wrapper.instance().getFoundTree(options, 'NODE1');
		const lowerFoundTree = wrapper.instance().getFoundTree(options, 'node1');

		expect(Object.keys(upperFoundTree).length).toEqual(Object.keys(lowerFoundTree).length);
	});
})
