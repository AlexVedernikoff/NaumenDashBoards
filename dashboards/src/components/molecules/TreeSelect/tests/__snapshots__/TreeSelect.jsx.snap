// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TreeSelect renders menu when "showMenu" property is changing 1`] = `
<div
  className="select"
>
  <Container
    className="valueContainer"
    disabled={false}
    onClick={[Function]}
  >
    <div
      className="value"
    >
      <Container
        className="label placeholder"
        title="Выберите значение"
      >
        Выберите значение
      </Container>
    </div>
  </Container>
</div>
`;

exports[`TreeSelect renders menu when "showMenu" property is changing 2`] = `
<div
  className="select"
>
  <Container
    className="valueContainer"
    disabled={false}
    onClick={[Function]}
  >
    <div
      className="value"
    >
      <Container
        className="label placeholder"
        title="Выберите значение"
      >
        Выберите значение
      </Container>
    </div>
  </Container>
  <OutsideClickDetector
    onClickOutside={[Function]}
  >
    <Container
      className="menu"
    >
      <SearchInput
        className=""
        focusOnMount={true}
        forwardedRef={
          Object {
            "current": null,
          }
        }
        onChange={[Function]}
        value=""
      />
      <Tree
        components={
          Object {
            "IndicatorsContainer": [Function],
            "LabelContainer": [Function],
            "MenuContainer": [Function],
            "Node": [Function],
            "SearchInput": [Function],
            "Tree": [Function],
            "ValueContainer": [Function],
          }
        }
        getNodeLabel={[Function]}
        getNodeValue={[Function]}
        getOptionLabel={[Function]}
        getOptionValue={[Function]}
        isDisabled={[Function]}
        loading={false}
        multiple={false}
        onSelect={[Function]}
        options={
          Object {
            "node0": Object {
              "children": Array [
                "node0$node0",
                "node0$node1",
              ],
              "id": "node0",
              "loading": false,
              "parent": null,
              "uploaded": true,
              "value": Object {
                "label": "node0",
                "value": "node0",
              },
            },
            "node0$node0": Object {
              "children": null,
              "id": "node0$node0",
              "loading": false,
              "parent": "node0",
              "uploaded": true,
              "value": Object {
                "label": "node0",
                "value": "node0",
              },
            },
            "node0$node1": Object {
              "children": null,
              "id": "node0$node1",
              "loading": false,
              "parent": "node0",
              "uploaded": true,
              "value": Object {
                "label": "node1",
                "value": "node1",
              },
            },
            "node1": Object {
              "children": Array [
                "node1$node0",
                "node1$node1",
              ],
              "id": "node1",
              "loading": false,
              "parent": null,
              "uploaded": true,
              "value": Object {
                "label": "node1",
                "value": "node1",
              },
            },
            "node1$node0": Object {
              "children": null,
              "id": "node1$node0",
              "loading": false,
              "parent": "node1",
              "uploaded": true,
              "value": Object {
                "label": "node0",
                "value": "node0",
              },
            },
            "node1$node1": Object {
              "children": null,
              "id": "node1$node1",
              "loading": false,
              "parent": "node1",
              "uploaded": true,
              "value": Object {
                "label": "node1",
                "value": "node1",
              },
            },
          }
        }
        searchValue=""
        showMore={false}
        value={null}
        values={Array []}
      />
    </Container>
  </OutsideClickDetector>
</div>
`;

exports[`TreeSelect renders when "removable" property is set to "true" 1`] = `
<div
  className="select"
>
  <Container
    className="valueContainer"
    disabled={false}
    onClick={[Function]}
  >
    <div
      className="value"
    >
      <Container
        className="label"
        title=""
      />
      <Container
        className="indicators"
        onClick={[Function]}
      >
        <IconButton
          active={false}
          className=""
          icon="REMOVE"
          onClick={[Function]}
          round={true}
          tip=""
          variant="INFO"
        />
      </Container>
    </div>
  </Container>
</div>
`;

exports[`TreeSelect renders when no options provided 1`] = `
<div
  className="select"
>
  <Container
    className="valueContainer"
    disabled={false}
    onClick={[Function]}
  >
    <div
      className="value"
    >
      <Container
        className="label placeholder"
        title="Выберите значение"
      >
        Выберите значение
      </Container>
    </div>
  </Container>
</div>
`;
